### ISSUE TITLE
Isolate workspaces into individual subscriptions, user resources into user resource groups, and service resources into a core resource group

### ISSUE THREAD

**Issue by cricex** [microsoft/AzureTRE#1073]
**Describe the solution you'd like**
As an Information Security Officer
I want each workspace deployed into their own spoke subscription, each user's resources deployed into their own resource group within the workspace's subscription, and service resources deployed into a service resource group within the workspace's subscription
So that we may improve security isolation, billing isolation, and overcome subscription resource limitations

![image](https://user-images.githubusercontent.com/7026048/141486727-4ed33e3c-ecaa-4eb9-b976-08fb8435f039.png)

**Comment by marrobi** [comment]
Thanks, we would be limited to 800 resource groups per subscription. Is that an issue? 

The separate resource groups for services and user resources can be handled in the respective templates.  So no changes to the TRE APIs, or resources processor are required.

The subscription piece is doable, but work required.

**Comment by cricex** [comment]
Is 800 a TRE or ARM limitation?  The documentation states that there is a limitation of 980 resource groups per subscription. (https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-subscription-service-limits#subscription-limits)

Regardless, I believe 800 or 980 resource groups (and therefore 800 or 980 individual researchers) is sufficiently high limitation for an individual workspace.

**Comment by Danny-Cooke** [comment]
This feature has been requested by a third party so will be looking at how to deliver this

**Comment by marrobi** [comment]
@SvenAelterman @Danny-Cooke is it worth us getting on a call to discuss requirements? Then a design can be posted here and discussed?

**Comment by SvenAelterman** [comment]
We can. Unfortunately, our project has not progressed.

**Comment by marrobi** [comment]
This branch shows how this might be done - haven't tested across multiple subscriptions.

https://github.com/marrobi/AzureTRE/tree/marrobi/workspace-different-subscription

Each workspace service would have to use the parent `workspace_subscription_id` when configuring the provider.

If someone wants to pick it up happy to help.

**Comment by TonyWildish-BH** [comment]
that looks interesting, thanks Marcus. We may be able to pick this up in the near future, we're mapping out our future strategy for scaling up, hence all the tickets recently.

**Comment by TonyWildish-BH** [comment]
> I am working with a customer who also needs a solution for 300+ workspaces. The customer is working with a partner on implementation.
> 
> I would propose we all work together on the solution? LMK if you'd be open to that.
> 
> If so, as a next step, we should individually review Marcus' work so far and then schedule a call to flesh out remaining work items and try to divide the tasks?

I'd be interested, thanks.

**Comment by SvenAelterman** [comment]
> This branch shows how this might be done - haven't tested across multiple subscriptions.
> 
> https://github.com/marrobi/AzureTRE/tree/marrobi/workspace-different-subscription
> 
> Each workspace service would have to use the parent `workspace_subscription_id` when configuring the provider.
> 
> If someone wants to pick it up happy to help.

We've evaluated this and this approach looks very good.

To extend this, we'd need to use a similar approach with an aliased azurerm provider for each workspace service, I presume? How would we get the subscription ID of the workspace passed on to the bundle for the workspace service without having to ask for it again in the UI?

I think we would use the same way the `workspace_id` gets passed along, but I can't recall how that happens? Is in the Python code or is it declaratively?

**Comment by marrobi** [comment]
Correct, could get the workspace subscription ID as discussed here - https://github.com/microsoft/AzureTRE/issues/4284

**Comment by marrobi** [comment]
@SvenAelterman have you managed to progress this at all?

**Comment by tamirkamara** [comment]
Thanks @SvenAelterman. 
In what you're building how & what will decide the subscription a workspace is created in?

**Comment by SvenAelterman** [comment]
@tamirkama The three current use cases will see the workspaces built by a research support team. The research support teams will enter a subscription ID as a parameter during workspace creation (just a text box). If the text box is blank, the workspace will be built in the same subscription as the core.

**Comment by TonyWildish-BH** [comment]
> @tamirkama The three current use cases will see the workspaces built by a research support team. The research support teams will enter a subscription ID as a parameter during workspace creation (just a text box). If the text box is blank, the workspace will be built in the same subscription as the core.

minor point, could we make it a subscription name, rather than an ID, for convenience.

**Comment by martinpeck** [comment]
I think @tamirkamara 's main question is "Is there a design for this?"

@SvenAelterman : is what you're describing being implemented by @askew (as he appears to have the PRs/commits related to this)?

**Comment by SvenAelterman** [comment]
> > @tamirkama The three current use cases will see the workspaces built by a research support team. The research support teams will enter a subscription ID as a parameter during workspace creation (just a text box). If the text box is blank, the workspace will be built in the same subscription as the core.
> 
> minor point, could we make it a subscription name, rather than an ID, for convenience.

I am not sure, but that would add extra complexity because the TF doesn't accept a subscription name in the provider config. For good reason, as subscription names are not unique.

**Comment by SvenAelterman** [comment]
> I think @tamirkamara 's main question is "Is there a design for this?"
> 
> @SvenAelterman : is what you're describing being implemented by @askew (as he appears to have the PRs/commits related to this)?

Yes, we are coordinating weekly.

**Comment by marrobi** [comment]
My view is that V1 can be subscription ID pasted into a box. It could also potentially be a hidden field by default - TBD.

Later on could have an API that returns subscriptions (that the MSI has access to) that populates a drop down, but don't think that's necessary for initial version.

**Comment by askew** [comment]
@tamirkamara,  
> 2. I understand each workspace will have its own subscription. Is that right? What will create it? How will the Resource Processor identity get permissions over the new subscriptions?

No, in this initial implementation, we are relying on the subscription already being created, and the resource processor managed identity having __Owner__ (or __Contributor__ & __User Access Administrator__) roles assignments on the subscription.

We are selecting the subscription simply by providing a subscription identifier as a parameter when creating the workspace.

The workspace subscription id will then be automatically passed to services and user resources created in the workspace.

We're not doing anything to limit the number of workspaces per subscription yet.

This should unblock a number of scenarios by simply allowing the TRE to deploy workspaces to a subscription separate to the core.

There are a number of nice-to-have enhancements that could be further added, but not necessary to get the basic feature working.

1. The UI could potentially query available subscriptions and have a drop-down list on names, rather than requiring a GUID (@TonyWildish-BH). It would still need to id passing to the bundle as that's what Terraform needs.
2. In very large scale scenarios, you could event limit the available subscription by the TRE user (admin) identity.
3. Subscription creation, or even just permission granting could be automated, but this would require a workload identity with permissions at the Management Group level.
4. Subscription suitability by capacity could be calculated by looking at quota usage for various resource type, i.e. storage accounts and VM cores.

Initially, though the implementation can be very straight forward.

**Comment by SvenAelterman** [comment]
On 3), permissions on the management group would not be sufficient to create subscriptions. The ID would also need permissions on the billing account (EA or MCA).
